lib/ansible/plugins/action/reboot.py

```diff
@@ -34,81 +34,130 @@ class ActionModule(ActionBase):
     DEFAULT_BOOT_TIME_COMMAND = 'cat /proc/sys/kernel/random/boot_id'
     DEFAULT_REBOOT_MESSAGE = 'Reboot initiated by Ansible'
     DEFAULT_SHUTDOWN_COMMAND = 'shutdown'
+    DEFAULT_SHUTDOWN_COMMAND_ARGS = '-r {delay_min} "{message}"'
     DEFAULT_SUDOABLE = True
 
     DEPRECATED_ARGS = {}
 
     BOOT_TIME_COMMANDS = {
-        'openbsd': '/sbin/sysctl kern.boottime',
         'freebsd': '/sbin/sysctl kern.boottime',
-        'sunos': 'who -b',
-        'darwin': 'who -b',
+        'openbsd': '/sbin/sysctl kern.boottime',
+        'macosx': 'who -b',
+        'solaris': 'who -b',
     }
 
     SHUTDOWN_COMMANDS = {
-        'linux': DEFAULT_SHUTDOWN_COMMAND,
-        'freebsd': DEFAULT_SHUTDOWN_COMMAND,
-        'openbsd': DEFAULT_SHUTDOWN_COMMAND,
-        'sunos': '/usr/sbin/shutdown',
-        'darwin': '/sbin/shutdown',
+        'alpine': 'reboot',
     }
 
     SHUTDOWN_COMMAND_ARGS = {
-        'linux': '-r {delay_min} "{message}"',
+        'alpine': '',
         'freebsd': '-r +{delay_sec}s "{message}"',
-        'sunos': '-i 6 -y -g {delay_sec} "{message}"',
-        'darwin': '-r +{delay_min} "{message}"',
+        'linux': DEFAULT_SHUTDOWN_COMMAND_ARGS,
+        'macosx': '-r +{delay_min} "{message}"',
         'openbsd': '-r +{delay_min} "{message}"',
+        'solaris': '-y -g {delay_sec} -r "{message}"',
+        'solaris10': '-y -g {delay_sec} -i 6 "{message}"',
     }
 
     def __init__(self, *args, **kwargs):
         super(ActionModule, self).__init__(*args, **kwargs)
 
         self._original_connection_timeout = None
         self._previous_boot_time = None
+        self._task_vars = None
+        self._distribution = None
+        self._shutdown_command = None
+
+    @property
+    def distribution(self):
+        if self._distribution is None:
+            self._distribution = dict()
+            try:
+
+                module_output = self._execute_module(
+                    task_vars=self._task_vars,
+                    module_name='setup',
+                    module_args=dict(
+                        filter='ansible_distribution*',
+                        gather_subset='min'))
+                if not module_output['ansible_facts'] or module_output['failed']:
+                    raise AnsibleError('Failed to determine system distribution.  {0}, {1}'.format(module_output['module_stdout'].strip(),
+                                                                                                   module_output['module_stderr'].strip()))
+            except KeyError as ke:
```

Comment:
This exception might be a little too tricky... If I understand it, it's handling the non-error case because you provoke KeyError if everything succeeded.  We shouldn't do that except in really special cases :-)

Reply:
yeah, or even simpler:
``` python
if 'failed' in module_output:
    raise AnsibleError([...])
```

containment checks (in) for dicts use the dict's keys already.